templates/default/pages/_app.jsx

```diff
@@ -13,6 +13,34 @@ import "../styles/docsearch.css";
 import "../styles/global.css";
 import "../styles/prism.css";
 
+// ToC: get the html nodelist for headings
+function collectHeadings(nodes) {
+  const sections = [];
+
+  Array.from(nodes).map((node) => {
+    // create toc with h2 and h3 elements
+    if (node.tagName === "H2" || node.tagName === "H3") {
+      const { id, innerText: title } = node;
+      if (id && title) {
+        if (node.tagName === "H3") {
+          if (sections[sections.length - 1]) {
+            return sections[sections.length - 1].children.push({
+              id,
+              title,
+            });
+          }
+        } else {
+          return sections.push({ id, title, children: [] });
+        }
+      }
+    }
+
+    return sections.push(...collectHeadings(node.children ?? []));
+  });
+
+  return sections;
+}
+
```

Comment:
What if we:
1. Replace `map` with `forEach`. This is more suitable for this use case as we don't need to create a new array, we just want to iterate over an array and do sth for some nodes. This way you also don't need to create artificial return statements in each if block.
2. Try to simplify the code a bit so that we don't have so many nested if blocks as it is really hard to reason about.

```suggestion
  Array.from(nodes).forEach((node) => {
    const { id, innerText: title, tagName: level } = node;
    if (!(id && title)) {
      return;
    }
    if (level === "H3") {
      const parentSection = sections[sections.length - 1];
      parentSection && parentSection.children.push({ id, title });
    } else if (level === "H2") {
      sections.push({ id, title, children: [] });
    }
    
    sections.push(...collectHeadings(node.children ?? []));
  });

  return sections;
}

```

Reply:
Have tested and this works. One minor change I had to add an if statement on line 44 as esbuild was throwing error during build `Expected an assignment or function call and instead saw an expression.`